Draft-Accepted/RFCXXXX-Command-Shell-Integration.md

```diff
@@ -0,0 +1,326 @@
+SupercededBy: NA
+Version: 1.0
+Area: Interactive Shell
+Comments Due: June 1, 2023
+Plan to implement: Yes
+---
+
+# Command Shell Integration
+
+Executables can provide a more integrated experience with a particular shell by indicating features it supports which the shell can leverage.
+
+Shell integration feature areas:
+
+- Structured output and formatting
+- Tab completion
+- Help
+- Predictors (PowerShell)
+- Feedback Providers (PowerShell)
+
+Traditionally, POSIX shells have pre-defined filesystem locations where tools would put their tab completion scripts or help content.
+However, this typically works for the system and not for a particular user who doesn't have administrative rights to the system.
+This can also make it more complicated to uninstall tools that place files across the filesystem.
+
+Although the examples here are focused on PowerShell, the design is intended to be generic and can be used by other shells.
+However, PowerShell specific implementation details are included.
+
+## Motivation
+
+As a tool developer,
+I can easily integrate with a shell,
+so that users are more productive having a consistent experience within that shell.
+
+## Current Experience
+
+Some tools that are executables and not modules/cmdlets have already reached out on how they can have better integration with PowerShell.
+The current experience requires the user to know about the integration scripts and add them to their `$profile` so that it gets loaded.
+This greatly decreases usage and thus motivation for tooling to produce integration scripts due to lack of discovery and also
+manual effort by the user.
+
+## Enhanced User Experience
+
+In these examples, we'll use a hypothetical version of `az` (Azure CLI) that implements all of the shell integration features.
+The user would simply install the tool and it would automatically be integrated with the shell.
+
+### Structured Output with Formatting
+
+```powershell
+az group list
+```
+
+Current output:
+
+```output
+[
+  {
+    "id": "/subscriptions/11111111-1111-1111-1111-111111111111/resourceGroups/RG1",
+    "location": "westcentralus",
+    "managedBy": null,
+    "name": "RG1",
+    "properties": {
+      "provisioningState": "Succeeded"
+    },
+    "tags": {
+      "RequestID": "1111"
+    },
+    "type": "Microsoft.Resources/resourceGroups"
+  },
+  {
+    "id": "/subscriptions/11111111-1111-1111-1111-111111111111/resourceGroups/RG2",
+    "location": "westcentralus",
+    "managedBy": null,
+    "name": "RG2",
+    "properties": {
+      "provisioningState": "Succeeded"
+    },
+    "tags": {
+      "RequestID": "1111"
+    },
+    "type": "Microsoft.Resources/resourceGroups"
+  }
+]
+```
+
+Although in PowerShell, this output can be piped to `ConvertFrom-Json`, it's an additional step for every `az` command execution.
+It would be preferrable to have the tool output structured data in a format that PowerShell can automatically convert to an object
+and used similarly to a cmdlet.
```

Comment:
This seems to pose some of the same problems of abstraction that the JSON adapter has? 

Are you proposing that PowerShell will just automatically convert any kind of JSON to objects or that the native command has to opt into such an experience?

Reply:
Customers should be able to enable/disable this behavior in the case of formatting issues or simply if they are re-using code that is not leveraging such a capability.

Reply:
I'm changing this to auto convert only JSON that includes `$typeNames` as it indicates it explicitly wants conversion/formatting.